File: example/rust/src/bin/5_fk_test.rs
use rebotarm_control_rt_rust_examples::common::{
    parse_floats, parse_urdf_path, print_pose_with_model, prompt, q_rad_from_deg, MathModel,
};
use std::env;
use std::error::Error;

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = env::args().skip(1).collect();
    let model = MathModel::load(&parse_urdf_path(&args))?;
    println!("Rust FK demo. Input six joint angles in degrees.");
    println!("Backend: C++/Pinocchio librebotarm_math.so");
    println!("examples: 0 0 0 0 0 0 | 45 -30 15 -60 90 180");
    loop {
        let Some(line) = prompt("joint angles > ")? else {
            break;
        };
        if matches!(line.as_str(), "q" | "quit" | "exit") {
            break;
        }
        let values = parse_floats(&line)?;
        if values.len() != 6 {
            println!("need exactly 6 joint angles");
            continue;
        }
        let q = q_rad_from_deg(&values);
        print_pose_with_model(&model, &q)?;
    }
    Ok(())
}

File: example/cpp/src/5_fk_test.cpp
#include "example_common.hpp"

#include <iostream>

int main(int argc, char** argv) {
  if (example::has_flag(argc, argv, "--help") || example::has_flag(argc, argv, "-h")) {
    std::cout << "Usage: ./5_fk_test [--urdf path/to/robot.urdf]\n";
    return 0;
  }

  rebotarm::RobotModel model(example::urdf_arg(argc, argv));
  std::cout << "C++ FK demo. Input " << model.nq() << " joint angles in degrees.\n";
  std::cout << "Backend: rebotarm::RobotModel + Pinocchio C++.\n";
  std::cout << "examples: 0 0 0 0 0 0 | 45 -30 15 -60 90 180\n";
  std::cout << "q / quit / exit to stop.\n\n";

  std::string line;
  while (true) {
    std::cout << "joint angles > ";
    if (!std::getline(std::cin, line)) break;
    if (line == "q" || line == "quit" || line == "exit") break;
    if (line.empty()) continue;
    try {
      const auto values = example::parse_numbers(line);
      const Eigen::VectorXd q = example::q_from_degrees(values, model.nq());
      example::print_pose(model, q);
    } catch (const std::exception& e) {
      std::cout << "invalid input: " << e.what() << "\n";
    }
  }
  return 0;
}
